src/compiler/compiler.rs

```diff
@@ -275,124 +296,205 @@ impl Compiler {
-            .collect::<HashMap<_, _>>();
-        // If `ForceRecache` is enabled, we won't check the cache.
-        let start = Instant::now();
-        let cache_status = if cache_control == CacheControl::ForceRecache {
-            Cache::Recache
-        } else {
-            storage.get(&key)
-        };
-        let duration = start.elapsed();
-        match cache_status {
-            Cache::Hit(mut entry) => {
-                debug!("[{}]: Cache hit!", out_file);
-                for (key, path) in &outputs {
-                    let mut f = try!(File::create(path));
-                    try!(entry.get_object(key, &mut f));
-                }
-                let mut stdout = io::Cursor::new(vec!());
-                let mut stderr = io::Cursor::new(vec!());
-                entry.get_object("stdout", &mut stdout).unwrap_or(());
-                entry.get_object("stderr", &mut stderr).unwrap_or(());
-                Ok((CompileResult::CacheHit(duration),
-                    process::Output {
-                        status: exit_status(0),
-                        stdout: stdout.into_inner(),
-                        stderr: stderr.into_inner(),
-                    }))
-            },
+        let result = self.kind.preprocess(creator, self, parsed_args, cwd, pool);
+        let out_file = out_file.into_owned();
+        let result = result.map_err(move |e| {
+            debug!("[{}]: preprocessor failed: {:?}", out_file, e);
+            e
+        });
 
-            res @ Cache::Miss | res @ Cache::Recache | res @ Cache::Error(_) => {
-                let miss_type = match res {
-                    Cache::Miss => { debug!("[{}]: Cache miss!", out_file); MissType::Normal }
-                    Cache::Recache => { debug!("[{}]: Cache recache!", out_file); MissType::ForcedRecache }
-                    Cache::Error(e) => {
-                        debug!("[{}]: Cache read error: {:?}", out_file, e);
+        let parsed_args = parsed_args.clone();
+        let cwd = cwd.to_string();
+        let arguments = arguments.to_vec();
+        let me = self.clone();
+        let storage = storage.clone();
+        let pool = pool.clone();
+        let creator = creator.clone();
+
+        Box::new(result.and_then(move |preprocessor_result| {
+            // If the preprocessor failed, just return that result.
+            if !preprocessor_result.status.success() {
+                debug!("[{}]: preprocessor returned error status {:?}",
+                       parsed_args.output_file(),
+                       preprocessor_result.status.code());
+                // Drop the stdout since it's the preprocessor output, just hand back stderr and the exit status.
+                let output = process::Output {
+                    stdout: vec!(),
+                    ..preprocessor_result
+                };
+                return Box::new(future::ok((CompileResult::Error, output))) as Box<Future<Item=_, Error=_>>
```

Comment:
Most of the time, yeah, but `.boxed()` has the added requirement of `Send` so here where most trait objects are not `Send` we can't always use it. In this case it inferred the closure return type as then a future which was `Send` causing errors for later returns.

There's other annotations that can happen here, though, so I'll apply those.